src/catchup/simulation/ApplyTransactionsWork.cpp

```diff
@@ -35,17 +49,190 @@ ApplyTransactionsWork::ApplyTransactionsWork(
     }
 }
 
+static void
+checkResults(Application& app, uint32_t ledger,
+             std::vector<TransactionResultPair> const& results)
+{
+    auto resSet = TransactionFrame::getTransactionHistoryResults(
+        app.getDatabase(), ledger);
+
+    assert(resSet.results.size() == results.size());
+    for (size_t i = 0; i < results.size(); i++)
+    {
+        auto const& res = results[i];
+        assert(res.transactionHash == resSet.results[i].transactionHash);
+
+        auto const& dbRes = resSet.results[i].result;
+        if (dbRes.result.code() != res.result.result.code())
+        {
+            auto msg = fmt::format(
+                "Expected result code {} does not agree with {} for tx {}",
+                res.result.result.code(), dbRes.result.code(),
+                binToHex(res.transactionHash));
+            CLOG(ERROR, "History") << msg;
+        }
+    }
+}
+
+bool
+ApplyTransactionsWork::hasSig(PublicKey const& account,
+                              TransactionEnvelope const& oldEnvelope,
+                              Hash const& hash)
+{
+    // Is the signature of this account present in the envelope we're
+    // simulating?
+    if (!mUsedPubKeys.insert(account).second)
+    {
+        return false;
+    }
+
+    for (auto const& sig : oldEnvelope.signatures)
+    {
+        if (PubKeyUtils::verifySig(account, sig.signature, hash))
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+void
+ApplyTransactionsWork::addSignerKeys(AccountID const& acc,
```

Comment:
this seems computationally intensive, do we need to do this because we need to "match" & replace signatures for partitions?

In any case, not a big deal if this is only done when we generate the stream (it would be if this was used while replaying).

Reply:
Yes, we need to make sure that all scaled transaction envelopes have correct signatures. It's unfortunate that this is pretty expensive, but we don't pay the price during the actual simulation (only during generation)